### PR TITLE
brand new cli tools written in rust

### PR DESCRIPTION
**Pull request by forehalo**
the working/tracing PR of refactoring napi cli tools with rust.

features: 
- [x] napi new  
- [x] napi build
- [x]  napi watch
- [ ] 🚘napi xxx

Aimed to fix all known issues with current cli as well:
- [x] #981
- [x] #988 
- [x] #407
- [x] #1061
- [x] #1170 
- [x] #1188 
- [ ] #1189 
- [ ] #1045

### CODE REVIEW

File: crates/cli/src/commands/build/mod.rs [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+      "CARGO_TARGET_{}_LINKER",
+      self.target.triple.to_uppercase().replace('-', "_")
+    );
+
+    let envs = [("CC", &cc), ("CXX", &cxx), (linker_env_name.as_str(), &cc)];
+    trace!("set envs for zig cross compiling");
+    for (name, value) in envs.iter() {
+      trace!("  {}: {}", &name, &value.display());
+      cmd.env(name, value);
+    }
```
Comment (messense) [inline crates/cli/src/commands/build/mod.rs]: The env vars are better aligned with `cargo-zigbuild`: https://github.com/messense/cargo-zigbuild/blob/9f31dcb97d5cff618d60b3c1d18dfb9f51045b7c/src/build.rs#L184-L233. 

And there are some hacks we do in `cargo-zigbuild` for some targets: https://github.com/messense/cargo-zigbuild/blob/9f31dcb97d5cff618d60b3c1d18dfb9f51045b7c/src/build.rs#L239-L287

Comment (forehalo) [reply crates/cli/src/commands/build/mod.rs]: The complexity of using `cargo-zigbuild` is much more than my expectation.

I would migrate the old logic from js version which makes it work first and well design the crossing compiling part later on.

Comment (messense) [reply crates/cli/src/commands/build/mod.rs]: Well, it's much easier if you do it this way: https://github.com/PyO3/maturin/blob/922459f19d5a1ba1295401f6b9c13198bcf61c2b/src/compile.rs#L202-L221 , all the env vars and operating system dependency hacks are taken care of.

File: crates/cli/index.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+const { run } = require('./index.node')
```
Comment (Brooooooklyn) [inline crates/cli/index.js]: Should we publish it as standalone binary?

Comment (forehalo) [reply crates/cli/index.js]: it exists only for test convenience

**Comment by forehalo** [comment]
Unfortunately, the rust cli implementation has to be deprecated because of the complexity of using.

We would soon introduce a redesigned cli in node with both command line and programatic interface.

**Pull request closed without merging**